### ISSUE TITLE
Underscore is forbidden as an alias for gettext

### ISSUE THREAD

**Issue by antlu** [wemake-services/wemake-python-styleguide#899]
# Bug report
## What's wrong
The use of `_` as an alias for **gettext** is considered a violation of the rule [WPS121](https://wemake-python-stylegui.de/en/latest/pages/usage/violations/naming.html#wemake_python_styleguide.violations.naming.UnusedVariableIsUsedViolation).

```python
from django.utils.translation import gettext as _

output = _("Welcome to my site.")
```

> Found usage of a variable marked as unused: _ flake8(WPS121)

## How is that should be
`_` should be allowed as an alias for **gettext()** and **gettext_lazy()**, as this is conventional:

> This function is usually aliased as _()

https://docs.python.org/3/library/gettext.html#gettext.gettext

> It’s convention to import this as a shorter alias, _, to save typing.

https://docs.djangoproject.com/en/2.2/topics/i18n/translation/#standard-translation

**Comment by sobolevn** [comment]
Well, yes, that's a bug! But, I am not sure how to fix it.

Let's ask our fellow experts: @lensvol @kxepal @orsinium

**Comment by lensvol** [comment]
Well, that one is interesting.

On the surface, we can use simple heuristic that looks for the import of the `gettext` (or `gettext_lazy` imported as an underscore. If we've had access to a corresponding AST, then it would be relatively easy.

But since we only have stream of tokens at our disposal, I would suggest looking for a sequence of token corresponding to `gettext as _` and use it as condition for raising a violation.

**Comment by sobolevn** [comment]
@lensvol we actually have `ast` access. But, I don't agree that checking import is a good idea. `flask-babel` uses the same technique. And there are libraries as well.

Maybe we can allow `_(STRING)` calls and `_(STRING.format(...))` as well? I guess this will cover this case perfectly.

**Comment by lensvol** [comment]
@sobolevn My bad, somehow missed that this is an `ASTViolation`. :)

Yes, your approach would be more appropriate here. The only problem that I see is that string can also be passed as a variable, how will we proceed in that case?

**Comment by sobolevn** [comment]
Good point. I don't know :)

In my projects I use `i18n` alias instead of `_`. But, it has some problems: sometimes messages are not collected because of this rename.

**Comment by sobolevn** [comment]
Maybe we can create a separate violation for `_()` case? And just silence it for `django` in the user's config.

**Comment by sobolevn** [comment]
It also conflicts with https://github.com/dry-python/lambdas

Solution: ignore this violation if this is a case for you.

**Comment by Dreamsorcerer** [comment]
Shouldn't this rule require that a variable has been defined somewhere?

e.g. There should be an `_ = ..` or `def f(_):` line somewhere before you label this as an unused variable?

I'm using gettext.install(), so `_` is never defined in my code, it's installed into the builtins. Therefore I see no reason it should be telling me that this variable has been marked as unused.

**Comment by sobolevn** [comment]
No, we do not require `_` to be defined anywhere.

**Comment by Dreamsorcerer** [comment]
I realise that's how it works now. But, it would resolve this issue if it ignored undefined variables (which get caught be a different rule, and get whitelisted with the builtins config option).

If a variable name `_` is not assigned anywhere or named as a parameter, then it's clearly not indicating an unused variable. If it has been imported from a module, or installed as a builtin, then it is explicitly being included for use.

**Comment by orsinium** [comment]
I agree with @Dreamsorcerer. We can forbid to define this variable (as a function argument or assignment) but allow us to use it when it is an import name or import alias. I've seen underscore usage as import name by-design in a lot of libs: short lambdas, pattern matching, strings wrapping. I don't think it's a good idea in the first place (because there are too many such libs, we can have name conflict) but it's definitely not users' fault but the lib devs.

Luckily, we already have a check to forbid `_ = something(...)` if I remain correctly. So, just allow using `_` in WPS121 should work well.

**Comment by sobolevn** [comment]
Oh, I see. Let's reopen it. I agree: using `_` is not a problem (because sadly that's how a lot of python developers force us to do), but defining `_` is a bad idea!

**Comment by orsinium** [comment]
Is it a bad construction? 

```python
prefix, _, suffix = some_str.partition('_')
```

Well, `_` can be replaced by `_sep` but I'm not sure is it ok or not to force users to do it.

**Comment by sobolevn** [comment]
yes, `_sep` is better, because it has context. In case you will need this variable your refactoring will look like `_sep -> sep`: it is clear and easy. On the other hand, `_ -> sep` requires you to know what is inside `_` which is an extra cognitive load.
